Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.cpp

```diff
@@ -125,56 +141,32 @@ std::unique_ptr<MediaRecorderPrivateWriter> MediaRecorderPrivateWriterWebM::crea
-    case kCMVideoCodecType_AV1: return mkvmuxer::Tracks::kAv1CodecId;
-    case kAudioFormatOpus: return mkvmuxer::Tracks::kOpusCodecId;
-    default:
-        ASSERT_NOT_REACHED("Unsupported codec");
-        return "";
-    }
-}
-
 MediaRecorderPrivateWriterWebM::MediaRecorderPrivateWriterWebM(MediaRecorderPrivateWriterListener& listener)
     : m_delegate(makeUniqueRef<MediaRecorderPrivateWriterWebMDelegate>(listener))
 {
 }
 
 MediaRecorderPrivateWriterWebM::~MediaRecorderPrivateWriterWebM() = default;
 
-std::optional<uint8_t> MediaRecorderPrivateWriterWebM::addAudioTrack(CMFormatDescriptionRef description)
+std::optional<uint8_t> MediaRecorderPrivateWriterWebM::addAudioTrack(const AudioInfo& description)
 {
-    const AudioStreamBasicDescription* asbd = PAL::CMAudioFormatDescriptionGetStreamBasicDescription(description);
-    if (!asbd)
-        return { };
-
-    return m_delegate->addAudioTrack(*asbd, mkvCodeIcForMediaVideoCodecId(asbd->mFormatID));
+    return m_delegate->addAudioTrack(description);
 }
 
-std::optional<uint8_t> MediaRecorderPrivateWriterWebM::addVideoTrack(CMFormatDescriptionRef description, const std::optional<CGAffineTransform>&)
+std::optional<uint8_t> MediaRecorderPrivateWriterWebM::addVideoTrack(const VideoInfo& description, const std::optional<CGAffineTransform>&)
 {
-    CMVideoDimensions dimensions = PAL::CMVideoFormatDescriptionGetDimensions(description);
-    CMVideoCodecType codec = PAL::CMFormatDescriptionGetMediaSubType(description);
-    return m_delegate->addVideoTrack(dimensions.width, dimensions.height, mkvCodeIcForMediaVideoCodecId(codec));
+    return m_delegate->addVideoTrack(description);
 }
 
-MediaRecorderPrivateWriterWebM::Result MediaRecorderPrivateWriterWebM::writeFrame(const MediaSample& sample)
+MediaRecorderPrivateWriterWebM::Result MediaRecorderPrivateWriterWebM::writeFrame(const MediaSamplesBlock& sample)
 {
-    RetainPtr sampleBuffer = downcast<MediaSampleAVFObjC>(sample).sampleBuffer();
-    RetainPtr buffer = PAL::CMSampleBufferGetDataBuffer(sampleBuffer.get());
-    ASSERT(PAL::CMBlockBufferIsRangeContiguous(buffer.get(), 0, 0));
-    size_t srcSize = PAL::CMBlockBufferGetDataLength(buffer.get());
-    char* srcData = nullptr;
-    if (PAL::CMBlockBufferGetDataPointer(buffer.get(), 0, nullptr, nullptr, &srcData) != noErr)
-        return Result::Failure;
-
-    bool isKeyFrame = sample.flags() & MediaSample::IsSync;
-
-    return m_delegate->addFrame({ reinterpret_cast<const uint8_t*>(srcData), srcSize }, sample.trackID(), Seconds { sample.presentationTime().toDouble() }.nanosecondsAs<uint64_t>(), isKeyFrame) ? Result::Success : Result::Failure;
+    bool success = true;
+    for (auto& block : sample) {
+        ASSERT(block.data);
+        Ref buffer = block.data->makeContiguous();
+        success &= m_delegate->addFrame(buffer->span(), sample.trackID(), Seconds { block.presentationTime.toDouble() }.nanosecondsAs<uint64_t>(), block.isSync());
```

Comment:
Does it make sense to call addFrame if the previous failed?

Reply:
Maybe not. But in this particular use with MediaRecorder a MediaSamplesBlock only ever contain a single MediaSampleItem

We never create a MediaSamplesBlock with more than two frames. 

Only MSE does so where we pack the audio samples in 2s buffer